### PR TITLE
Introduction of serialization instances in support of ledger peer snapshot

### PR DESCRIPTION
**Pull request by crocodile-dentist**
# Description

This change adds support for serialization of ledger peers by defining instances for its subcomponent types. The snapshot type will be introduced in a future commit.

# Checklist

- Branch
    - [x] Updated changelog files.
    - [ ] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [ ] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (coot) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:88]: This PR should also include a test that `fromCBOR` and `toCBOR` are compatible, e.g.
```hs
 fromCBOR . toCBOR = id
```

But also check that the `toCBOR` produces valid CBOR.  This is especially important if we don't have CDDL spec for this definition.   This can be done by encoding and then decoding with [`decodeTermToken`](https://hackage.haskell.org/package/cborg-0.2.10.0/docs/Codec-CBOR-FlatTerm.html#v:decodeTermToken) and validating with ` [`validFlatTerm`](https://hackage.haskell.org/package/cborg-0.2.10.0/docs/Codec-CBOR-FlatTerm.html#v:validFlatTerm); check `prop_codec_valid_cbor_encoding`.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (coot) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:87]: Two things:

* `case` expressions are a better fit (one extension less to use)
* We need a `fail` case so one will get a nice exception when the wrong data is received.

```suggestion
    case mode of
      0 -> do
         domain <- fromCBOR
         return $ RelayAccessDomain domain port
      1 -> do
         ip4 <- IP.IPv4 . IP.toIPv4 <$> fromCBOR
         return $ RelayAccessAddress ip4 port
      2 -> do
         ip6 <- IP.IPv6 . IP.toIPv6 <$> fromCBOR
         return $ RelayAccessAddress ip6 port
      a -> do
         fail ... 
```

Comment (coot) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:87]: Two things:

* `case` expressions are a better fit (one extension less to use)
* We need a `fail` case so one will get a nice exception when the wrong data is received.

```suggestion
    tag <- decodeWord8
    port <- fromInteger <$> fromCBOR @Integer 
    case tag of
      0 -> do
         domain <- fromCBOR
         return $ RelayAccessDomain domain port
      1 -> do
         ip4 <- IP.IPv4 . IP.toIPv4 <$> fromCBOR
         return $ RelayAccessAddress ip4 port
      2 -> do
         ip6 <- IP.IPv6 . IP.toIPv6 <$> fromCBOR
         return $ RelayAccessAddress ip6 port
      _ -> do
         fail ("RelayAccessPoint decoder: invalid tag " ++ show tag)
```

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (coot) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:71]: Isn't it a bit odd to encode the port before the IP address, domain name? We're all used to `IP:PORT`s.  Although, it's binary encoding so maybe it doesn't really matter.

**Review by coot — CHANGES_REQUESTED** [review]
Very nice, a few comments and a request to write some tests.

File: ouroboros-network-api/CHANGELOG.md
Comment (crocodile-dentist) [reply ouroboros-network-api/CHANGELOG.md:13]: Yes, consensus answers all queries with a toCBOR eventually.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:71]: Yes, but when port comes first, the FromCBOR instance looks slightly more elegant I think.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:87]: You're right, I've missed this.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:65]: Unfortunately, I think that's the case indeed.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:94]: Done.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (bolt12) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:101]: Is this according to the current CDDL spec?

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (bolt12) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:48]: I think it would be better to write the instance by hand and document it on the CDDL spec.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (bolt12) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:101]: Also CBOR codecs in ouroboros-network-protocols where the cddl spec is defined and the codecs are defined use:

```
import Codec.CBOR.Decoding qualified as CBOR
import Codec.CBOR.Encoding qualified as CBOR
```

and the `Codec` type. Take a look at how things are done there. I am not sure about defining the CDDL, though because this project doesn't deal with it... hmm

**Review by bolt12 — CHANGES_REQUESTED** [review]

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:101]: I haven't seen the CDDL spec for this, but I could create one if this is a requirement. I guess we should establish who should own this spec in the first place, us, node or consensus?

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:48]: One wrinkle here is that is you look at [Slot](https://github.com/IntersectMBO/cardano-base/blob/master/cardano-slotting/src/Cardano/Slotting/Slot.hs#L40) in cardano-slotting, the SlotNo type, which is a component of the snapshot, has a newtype automatically derived instance for Serialise, which is leveraged by the ToCBOR instance. Same for WithOrigin. And so I might as well bite the bullet and keep my auto derived instances as well I think.

Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:48]: So I've issued a PR to cardano-base so that hand written instances can be made for the WithOrigin type for recording the slot number of the snapshot.

Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:48]: So I've hand written all instances necessary to serialise the snapshot.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs]: The path to writing a spec is opened now that all serialisation instances are hand written.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (bolt12) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:114]: Why do you need OVERLAPPING pragma?

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (bolt12) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:130]:
```suggestion
instance {-# OVERLAPPING #-} FromCBOR (WithOrigin SlotNo) where
  fromCBOR = do
    listLen <- Codec.decodeListLen
    tag <- Codec.decodeWord8
    case (listLen, tag) of
      (1, 0) -> pure Origin
      (1, _) -> fail "LedgerPeers.Type: Expected tag for Origin constructor"
      (2, 1) -> At . SlotNo <$> fromCBOR
      (2, _) -> fail "LedgerPeers.Type: Expected tag for At constructor"
      _      -> fail "LedgerPeers.Type: Unrecognized list length while decoding WithOrigin SlotNo"
```

**Review by bolt12 — APPROVED** [review]
Looks good to me! I just left some small comments

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:114]: There is a ```ToCBOR (WithOrigin a)``` instance in cardano-slotting - I've made a patch PR to make it OVERLAPPABLE there. I'll hold off on merging this one until it gets approved. If not, I guess I can just create a newtype wrapper.

Comment (bolt12) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:114]: If you already have an instance for `WithOrigin` than it is not our responsibility to provide one since that's done in a different place. We should provide instance for types we own, so in this case we can't do much the way I see it! Sorry if I was misleading regarding CBOR instances, but I don't think we should go through the trouble of making these overlappable. They provide the `WithOrigin` instance and we just either need to provide the `SlotNo` instance or get it from somewhere if we don't own that type

Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs]: Their definition looks like this:
`instance {-# OVERLAPPABLE #-} (Serialise t, Typeable t) => ToCBOR (WithOrigin t) where
  toCBOR = encode`
I guess to have full control of the encoding this is not really useable. I've created a newtype wrapper instead.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (coot) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:65]: Do we have a round trip tests for `{To,From}CBOR` instances?

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (coot) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:135]: Do we have round trip tests for `{To,From}CBOR` instances?

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (coot) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:71]: Do we have round trip tests for `{To,From}JSON` instances?

**Review by coot — CHANGES_REQUESTED** [review]

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/RelayAccessPoint.hs:65]: Yes.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:135]: Yes.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:71]: Yes.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:142]: It encodes it as a two element list of Int's: the numerator and denominator.

Comment (coot) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:142]: Let's document the encoding in the `LedgerPeerSnapshot` type.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:181]: Done

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:188]: Done

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:135]: There's a test that checks if:
1. CBOR encoding is valid
2. CBOR round trip = identity
3. JSON decodes with the following stipulations:

   - since I've fully qualified every domain name, round trip equality test is modulo this
   - since serialisation of stake is in floating point, comparison to original Rational values is approximate

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:71]: Yes, see above.

File: ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs
Comment (coot) [inline ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:251]: So this is only to not parse `1.1.1.1.` as an `IP` address.  For that we'd also need a newer version of `iproute`.

I  would avoid this at all. This should be in the ledger rules.

Comment (crocodile-dentist) [reply ouroboros-network-api/src/Ouroboros/Network/PeerSelection/LedgerPeers/Type.hs:251]: No, IP addresses are handled by the domain constructor. This is a check to ensure that domain name we have read is fully qualified, and if not, add a period at the end. I realize now that this is redundant since I serialise all domain names in fully qualified format - maybe I wrote the former before I added the latter.

**Review by coot — APPROVED** [review]
LGTM

**Pull request merged**
